chain/epoch-manager/src/lib.rs

```diff
@@ -675,10 +677,24 @@ impl EpochManager {
                 epoch_config.max_inflation_rate,
             )
         };
-        let next_next_epoch_config = self.config.for_protocol_version(next_next_epoch_version);
-        let next_epoch_version = next_epoch_info.protocol_version();
-        let next_shard_layout = self.config.for_protocol_version(next_epoch_version).shard_layout;
-        let has_same_shard_layout = next_shard_layout == next_next_epoch_config.shard_layout;
+        let mut next_next_epoch_config = self.config.for_protocol_version(next_next_epoch_version);
+        let next_shard_layout = match next_epoch_info.shard_layout() {
+            // With dynamic resharding enabled, shard layout is stored in EpochInfo
+            Some(layout) => layout.clone(),
+            // Otherwise, fall back to the layout defined in config (this is needed both for
+            // compatibility and bootstrapping)
+            None => {
+                let next_protocol_version = next_epoch_info.protocol_version();
+                self.config.for_protocol_version(next_protocol_version).shard_layout
+            }
+        };
+
+        if ProtocolFeature::DynamicResharding.enabled(next_next_epoch_version) {
+            next_next_epoch_config.shard_layout = next_shard_layout.clone();
```

Comment:
Here you overwrite the epoch config. I originally assumed that you will make the epoch info the source of truth for the shard layout. How do those approaches differ? Why did you chose this one?